File: OpenSesameServer/openSesameServer.py
import tornado.httpserver
import tornado.websocket
import tornado.ioloop
import tornado.web

import uuid
import logging
import re
import json

STARTUP_TYPE = "startup"
REQUEST_TYPE = "request"
ACK_TYPE = "ack"
MISC_MSG_TYPE = "miscMsg"
DEBUG_TYPE = "debug"

#JSONification of Messages
# createJSONMsg
# idval: UIUD value given to session
# typeval: type of message
# message: any misc. message
def createJSONMsg(idval, typeval, message):
  retVal =  { "id" : idval, "type" : typeval , "message" : message }
  return json.dumps(retVal)
# Note: json.loads is decoder

class WSHandler(tornado.websocket.WebSocketHandler):
  clients = {}

  def check_origin(self, origin):
    return True

  # removeKey
  # Completely Removes the key-value pair from the map
  # Params
  # d: map to remove key-value pair from 
  # key: the key to remove
  def removeKey(d, key):
    r = dict(d)
    del r[key]
    return r

  def open(self):
    self.id = uuid.uuid4()
    self.write_message(createJSONMsg(None, MISC_MSG_TYPE, "You are now connected to the OpenSesame Server. You are Client "+self.id.hex))
    for key in self.clients:
      self.clients[key].write_message(createJSONMsg(self.id.hex, MISC_MSG_TYPE, "A new client has enetered the server"))
    self.clients[self.id.hex] = self
    logging.info('Connection '+self.id.hex+' was opened')

  def on_message(self, message):
    # So... is this acceptable? 
    def sendMsgWrapper(session_to,session_from,typeval,message):
      buildMsg = ""
      sessionFrom = ""
      if typeval == MISC_MSG_TYPE and not session_from == None:
        buildMsg = session_from.id.hex + ": " + message
      	sessionFrom = session_from.id.hex
      elif typeval == REQUEST_TYPE and not session_from == None:
      	buildMsg = message
      	sessionFrom = session_from.id.hex
      else:
        buildMsg = message
      	sessionFrom = None
      if session_to == None:
      	logging.error("session_to was NoneType")
      	if not session_from == None:
      	  session_from.write_message(createJSONMsg(None, MISC_MSG_TYPE, "session_to was NoneType"))
      else:
      	logging.debug("session_to: "+session_to.id.hex)
      	session_to.write_message(createJSONMsg(sessionFrom, typeval, buildMsg))

    global target
    logging.info('Incoming message:' +  message)

    decodedMsg = json.loads(message)
    if decodedMsg["type"] == STARTUP_TYPE:
      target = self
      logging.info("Found the target, Client "+target.id.hex)
    elif decodedMsg["type"] == REQUEST_TYPE:
      sendMsgWrapper(target,self,REQUEST_TYPE,"")
    elif decodedMsg["type"] == ACK_TYPE:
      logging.info('Client ' + self.id.hex + ' has achieved interaction with the Target')
      sendMsgWrapper(self.clients[decodedMsg["id"]], None, MISC_MSG_TYPE, "The target has received your message")
    elif decodedMsg["type"] == MISC_MSG_TYPE:
      for key in self.clients:
        if not target == None and not self.clients[key].id.hex == target.id.hex and not self.id.hex == target.id.hex:
          sendMsgWrapper(self.clients[key], self, MISC_MSG_TYPE, decodedMsg["message"])
    elif decodedMsg["type"] == DEBUG_TYPE:
      sendMsgWrapper(target,self,DEBUG_TYPE,"Response ping from OpenSesame Server")

  def on_close(self):
    logging.info('Connection '+self.id.hex+' was closed...')
    del self.clients[self.id.hex]
    for key in self.clients:
      self.clients[key].write_message(createJSONMsg(None, MISC_MSG_TYPE, "A client has left the server"))

# application_ssl = tornado.web.Application([
#   (r'/ws', WSHandler),
# ])

application = tornado.web.Application([
  (r'/ws', WSHandler),
])

if __name__ == "__main__":
  global target
  target = None
  logging.basicConfig(filename='pysocketLog.log', filemode='w', format='%(asctime)s %(name)-12s %(levelname)-8s %(message)s', level=logging.DEBUG)

#   ssl_server = tornado.httpserver.HTTPServer(application, ssl_options={
#     "certfile": "test.crt",
#     "keyfile": "test.key",
#   })
#   ssl_server.listen(443)

# Mobile (Cellular) networks seem to block/disable/mess around with a bunch of ports. 
# 80 and 443 are the only one I found that worked. The websocket sessions can be encrypted with
# SSL, but the "untrusted certificate" warning is a major inconvenience for this purpose
  http_server = tornado.httpserver.HTTPServer(application)
  http_server.listen(443)
  tornado.ioloop.IOLoop.instance().start()

File: OpenSesameServer/nodejs/openSesameServer.js
// Open Sesame Node.js Version
//

var app = require('express')();
var httpServer = require('http').Server(app);
var io = require('socket.io')(httpServer, { pingTimeout: 60000 });

// MESSAGE TYPES
// STARTUP_TYPE: Message Type that target device will send to open session
// REQUEST_TYPE: Message Type that triggers target device action
// ACK_TYPE: Message Type that target device sends back that confirms request fulfillment
// MISC_MSG_TYPE: Message Type for anything else
// DEBUG_TYPE: Message Type that was used for pinging between server and target device
// ERROR_MSG_TYPE: Message Type for error messages
var STARTUP_TYPE = "startup";
var REQUEST_TYPE = "request";
var ACK_TYPE = "ack";
var MISC_MSG_TYPE = "miscMsg";
var DEBUG_TYPE = "debug";
var ERROR_MSG_TYPE = "errorMsg";
var DEVICE_ASK_TYPE = "areYouDevice";

var targetDeviceSocket = {};

// var requestRoom = "requestRoom";

app.get('/', function(req,res) {
    res.sendFile(__dirname + '/index.html');
});

function getDeviceId(socket) {
    for(var key in targetDeviceSocket) {
        if(socket === targetDeviceSocket[key]) {
            return key;
        }
    }
    return null;
}

io.on('connection', function(socket) {
    console.log('a user connected');
    // A cry for connection
    socket.emit(DEVICE_ASK_TYPE, " ");

    socket.on('disconnect', function() {
        console.log('user disconnected');
        var targetId = getDeviceId(socket);
        if(targetId !== null) {
            console.log('Target device ' + targetId + ' has disconnected.');
            targetDeviceSocket[targetId] = null;
        }
    });

    socket.on(STARTUP_TYPE, function(msg){
        console.log('Target Device ' + msg + ' has connected');
        targetDeviceSocket[msg] = socket;
    });

    socket.on(MISC_MSG_TYPE, function(msg){
        console.log('message: ' + msg);
        io.emit(MISC_MSG_TYPE, msg);
    });

    socket.on(REQUEST_TYPE, function(msg){
        console.log('request received from client, forwarding request to target...');
        // send message only to the target device
        if(msg in targetDeviceSocket && targetDeviceSocket[msg] !== null) {
            targetDeviceSocket[msg].emit(REQUEST_TYPE, socket.id);
            // socket.join(requestRoom);
        } else {
            console.log('Target device ' + msg + ' is currently not connected');
            socket.emit(ERROR_MSG_TYPE,'Target device ' + msg + ' is currently not connected');
        }
    });

    socket.on(ACK_TYPE, function(msg){
        console.log('request fulfillment received from target device. returning acknowledgement...');
        io.to(msg).emit(MISC_MSG_TYPE,'request fulfillment received from target device. returning acknowledgement...');
    });

});

httpServer.listen(3000, function() {
    console.log('listening on *:3000');
});



